klog.go

```diff
@@ -793,23 +775,54 @@ func (l *loggingT) output(s severity, buf *buffer, file string, line int, alsoTo
 		// Write the stack trace for all goroutines to the files.
 		trace := stacks(true)
 		logExitFunc = func(error) {} // If we get a write error, we'll still exit below.
-		for log := fatalLog; log >= infoLog; log-- {
-			if f := l.file[log]; f != nil { // Can be nil if -logtostderr is set.
-				f.Write(trace)
-			}
-		}
+		l.writeLogData(fatalLog, trace)
 		l.mu.Unlock()
 		timeoutFlush(10 * time.Second)
 		os.Exit(255) // C++ uses -1, which is silly because it's anded with 255 anyway.
 	}
 	l.putBuffer(buf)
 	l.mu.Unlock()
+	// Note: The stats estimate logs for each severity level individually,
+	// even in the situation that log-file is specified and
+	// all severity-level logs are tracked only in the infoLog file.
 	if stats := severityStats[s]; stats != nil {
 		atomic.AddInt64(&stats.lines, 1)
 		atomic.AddInt64(&stats.bytes, int64(len(data)))
 	}
 }
 
+// writeLogData writes |data| to the `s` and lower severity files.
+// e.g. If Severity level is Error, the data will be written to all the Error,
+// Warning, and Info log file. However, if --log_file flag is provided, klog
+// no longer tracks logs separately due to their severity level, but rather
+// only write to the info logfile which later on will be flushed to the
+// dedicated log_file.
+func (l *loggingT) writeLogData(s severity, data []byte) {
+	if l.logFile != "" { // single-file mode
+		if l.file[infoLog] == nil {
```

Comment:
Can we use a dedicated buffer for this, rather than repurposing infoLog?

Reply:
Agree we shouldn't repurpose infolog. But I also have some concerns to use a dedicated buffer:
I want to patch a quick/safe fix to the log duplicate problem and catch the 1.15 release. Defining a new buffer results in much more complicated changes and it should have longer testing time before imported to kubernetes/kubernetes.

Updated the code with a dedicated buffer for single mode. Let me know if that's the reasonable approach and is safe to go in. If not, I'd prefer a quick fix that touches the least code.